File: 2020/rust/aoc/src/day5.rs
fn get_seat_id(item: &str) -> usize {
    item.chars().fold(0, |acc, elem| match elem {
        'R' => (acc << 1) + 1,
        'L' => acc << 1,
        'B' => (acc << 1) + 1,
        'F' => acc << 1,
        _ => panic!("Unknown letter {}", elem),
    })
}

fn part1(data: &[String]) -> usize {
    data.iter().map(|x| get_seat_id(x)).max().unwrap()
}

fn part2(data: &[String]) -> usize {
    let sorted = {
        let mut unsorted: Vec<_> = data.iter().map(|x| get_seat_id(x)).collect();
        unsorted.sort();
        unsorted
    };
    sorted[1..]
        .iter()
        .zip(sorted[..sorted.len() - 1].iter())
        .map(|(r, l)| r - l)
        .enumerate()
        .find(|(_, val)| *val == 2)
        .map(|(idx, _)| idx)
        .unwrap()
        + sorted[0]
        + 1
}

#[cfg(test)]
mod tests {

    #[test]
    fn day5_part1() {
        let data = [
            "FBFBBFFRLR".to_string(),
            "BFFFBBFRRR".to_string(),
            "FFFBBBFRRR".to_string(),
            "BBFFBBFRLL".to_string(),
        ];
        let truths = [357, 567, 119, 820];
        for (item, truth) in data.iter().zip(&truths) {
            assert_eq!(*truth, crate::day5::get_seat_id(item));
        }
        assert_eq!(820, crate::day5::part1(&data));
        assert_eq!(880, crate::day5::part1(&crate::parsers::day5()));
    }

    #[test]
    fn day5_part2() {
        assert_eq!(731, crate::day5::part2(&crate::parsers::day5()));
    }
}

File: 2020/fsharp/AOC/Day5.fs
module AOC.Day5

open System

let getSeatID item = 
    item 
    |> Seq.map (fun l -> 
        match l with 
        | 'L' -> '0'
        | 'R' -> '1'
        | 'B' -> '1'
        | 'F' -> '0'
        | _ -> l)
    |> Seq.toArray
    |> System.String
    |> (+) "0b"
    |> int

// 874.601 us
let part1 data = 
    data |> Seq.map getSeatID |> Seq.max

// 892.714 us
let part2 data =
    let sorted = data |> Seq.map getSeatID |> Array.ofSeq |> Array.sort
    Array.map2 (-) sorted.[1..] sorted.[..((Array.length sorted) - 2)]
    |> Array.findIndex ((=) 2)
    |> (+) sorted.[0]
    |> (+) 1 // 0 based indexing offset